src/ripple/app/misc/impl/TxQ.cpp

```diff
@@ -83,15 +83,16 @@ TxQ::FeeMetrics::update(Application& app,
     TxQ::Setup const& setup)
 {
     std::vector<uint64_t> feeLevels;
-    feeLevels.reserve(txnsExpected_);
+    auto const size = std::distance(view.txs.begin(), view.txs.end());
```

Comment:
So `txs` is of type `detail::ReadViewFwdRange<...>` which marks its iterators with `std::forward_iterator_tag`. As a result, `std::distance` has linear complexity here. Mildly annoying and enough for me to call it out, but not enough to ask that we look at this more closely. 

But there's something worse here, hidden under some layers of complexity: `view.txs.begin` calls `ReadView::txsBegin` and `view.txs.end` calls `ReadView::txsEnd` which themselves invoke `std::make_unique`; so this single line ends up doing two heap allocations too. Those are in addition to the two we do in the ranged `for` loop below. :disappointed: 

Considering that, I think it's a good idea to at least have you look at this a little more closely and see if there's, maybe, a way for us to:
1) directly get the size from the underlying container without making the code too ugly; and
2) avoid the dynamic allocation for the iterators in the ranged for.

Reply:
I wasn't too worried about the O(n) because it's still gonna be faster than `vector::reserve(maxint)`. :grinning: However, I agree about not liking those allocations. Maybe we can find a way to cache / memoize the iterators...

I agree with @seelabs about not adding an API until we can do this correctly. But the problem with "some reasonable amount" is figuring out what is reasonable. I thought `txnsExpected_` was reasonable until we had problems with it in the performance testing.  And since we're worrying about allocations, my guess is that the allocation cost of making it too small will be bad, and the time cost of making it too large is going to be worse than the linear time count of all the elements. But those are just guesses. We'd have to experiment and do more performance testing to be sure.

Speaking of performance testing, @mtrippled confirmed that with this patch, the performance test results are right in the normal range, so whatever inefficiencies are introduced, they can't be _too_ bad.

Reply:
:+1:

Reply:
@nbougalis  @ximinez : I took a look at this. Here's my thoughts:

The problem we have is we are iterating though a "virtual" collection of elements, not a concrete collection like "map". As a result:

1) We use pointers to base classes as iterators. This requires an allocation for the iterator and a virtual function call to both increment the iterator and dereference the iterator.

2) Not all collections hold value type we need, and the value type needs to be created. When this is done, it requires both parsing a serialization, and more allocations. Here's an example from `OpenView.cpp`
```
value_type
    dereference() const override
    {
        value_type result;
        {
            SerialIter sit(
                iter_->second.first->slice());
            result.first = std::make_shared<
                STTx const>(sit);
        }
        if (metadata_)
        {
            SerialIter sit(
                iter_->second.second->slice());
            result.second = std::make_shared<
                STObject const>(sit, sfMetadata);
        }
        return result;
    }
};
```

I can solve issue (1) by defining a virtual `forEachTxn` on each of the view classes. This will take a `std::function<void(tx_type const&)>` as a parameter. Unfortunately, for lambdas that capture state, this may allocate. There are solutions to this (a `function_ref` class that doesn't take ownership of the callable. One is proposed for standardization here: http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2018/p0792r3.html. Facebook's folly library also an implementation we could start with. Of course, we have team members that can write this as well. We can also use `std::reference_wrapper`, but that is cumbersome and I wouldn't expect callers to do that). In practice, even if it does allocate it will still be faster than the existing solution - one allocation vs two, and one virtual function call vs. N. And in many cases it won't alloc. I have a patch that shows this solution here: https://github.com/seelabs/rippled/commit/f1028c2f39b452b4a9b5e68907144dd874ab9d87

I don't see a good solution to (2). Given the allocations while iterating, optimizing the initial extra allocations for the iterators and the extra virtual function calls aren't going to matter. They will be swamped by the other allocations.